File: JaPetnese-Android/app/src/main/java/com/japetnese/app/formatter/JapaneseTimeFormatter.kt
package com.japetnese.app.formatter

import java.util.Calendar
import java.util.Date

object JapaneseTimeFormatter {
    private val dayOfWeekNames = arrayOf("日", "月", "火", "水", "木", "金", "土")
    private val amPmNames = arrayOf("午前", "午後")
    private val amPmReadings = arrayOf("ごぜん", "ごご")

    private val hourReadings = mapOf(
        1 to "いちじ", 2 to "にじ", 3 to "さんじ", 4 to "よじ",
        5 to "ごじ", 6 to "ろくじ", 7 to "しちじ", 8 to "はちじ",
        9 to "くじ", 10 to "じゅうじ", 11 to "じゅういちじ", 12 to "じゅうにじ"
    )

    private val monthReadings = mapOf(
        1 to "いちがつ", 2 to "にがつ", 3 to "さんがつ", 4 to "しがつ",
        5 to "ごがつ", 6 to "ろくがつ", 7 to "しちがつ", 8 to "はちがつ",
        9 to "くがつ", 10 to "じゅうがつ", 11 to "じゅういちがつ", 12 to "じゅうにがつ"
    )

    private fun cal(date: Date = Date()): Calendar = Calendar.getInstance().apply { time = date }

    fun formatHour(date: Date = Date()): String {
        val h = cal(date).get(Calendar.HOUR_OF_DAY)
        val dh = if (h == 0) 12 else if (h > 12) h - 12 else h
        return "${dh}時"
    }

    fun formatMinute(date: Date = Date()): String {
        val m = cal(date).get(Calendar.MINUTE)
        return if (m == 0) "" else "${m}分"
    }

    fun formatAmPm(date: Date = Date()): String {
        val h = cal(date).get(Calendar.HOUR_OF_DAY)
        return if (h < 12) amPmNames[0] else amPmNames[1]
    }

    fun formatTime(date: Date = Date()): String {
        val amPm = formatAmPm(date)
        val hour = formatHour(date)
        val min = formatMinute(date)
        return "$amPm$hour$min"
    }

    fun formatDate(date: Date = Date()): String {
        val c = cal(date)
        val month = c.get(Calendar.MONTH) + 1
        val day = c.get(Calendar.DAY_OF_MONTH)
        val weekday = c.get(Calendar.DAY_OF_WEEK) - 1 // Sunday=0
        return "${month}月${day}日 ${dayOfWeekNames[weekday]}曜日"
    }

    fun formatDateShort(date: Date = Date()): String {
        val c = cal(date)
        return "${c.get(Calendar.MONTH) + 1}月${c.get(Calendar.DAY_OF_MONTH)}日"
    }

    // Hiragana
    fun formatAmPmHiragana(date: Date = Date()): String {
        return if (cal(date).get(Calendar.HOUR_OF_DAY) < 12) amPmReadings[0] else amPmReadings[1]
    }

    fun formatHourHiragana(date: Date = Date()): String {
        val h = cal(date).get(Calendar.HOUR_OF_DAY)
        val dh = if (h == 0) 12 else if (h > 12) h - 12 else h
        return hourReadings[dh] ?: "${dh}じ"
    }

    fun formatMinuteHiragana(date: Date = Date()): String {
        val m = cal(date).get(Calendar.MINUTE)
        if (m == 0) return ""
        return minuteFullReading(m)
    }

    fun formatTimeShortHiragana(date: Date = Date()): String {
        return formatHourHiragana(date) + formatMinuteHiragana(date)
    }

    fun formatDateHiragana(date: Date = Date()): String {
        val c = cal(date)
        val month = c.get(Calendar.MONTH) + 1
        val day = c.get(Calendar.DAY_OF_MONTH)
        val weekday = c.get(Calendar.DAY_OF_WEEK) - 1
        val dayReadings = arrayOf("にち", "げつ", "か", "すい", "もく", "きん", "ど")
        return "${monthReadings[month]} ${dayReading(day)} ${dayReadings[weekday]}ようび"
    }

    private fun minuteFullReading(m: Int): String {
        val base = mapOf(
            1 to "いっ", 2 to "に", 3 to "さん", 4 to "よん", 5 to "ご",
            6 to "ろっ", 7 to "なな", 8 to "はっ", 9 to "きゅう", 10 to "じゅっ",
            11 to "じゅういっ", 12 to "じゅうに", 13 to "じゅうさん", 14 to "じゅうよん",
            15 to "じゅうご", 16 to "じゅうろっ", 17 to "じゅうなな", 18 to "じゅうはっ",
            19 to "じゅうきゅう", 20 to "にじゅっ",
            21 to "にじゅういっ", 22 to "にじゅうに", 23 to "にじゅうさん", 24 to "にじゅうよん",
            25 to "にじゅうご", 26 to "にじゅうろっ", 27 to "にじゅうなな", 28 to "にじゅうはっ",
            29 to "にじゅうきゅう", 30 to "さんじゅっ",
            31 to "さんじゅういっ", 32 to "さんじゅうに", 33 to "さんじゅうさん", 34 to "さんじゅうよん",
            35 to "さんじゅうご", 36 to "さんじゅうろっ", 37 to "さんじゅうなな", 38 to "さんじゅうはっ",
            39 to "さんじゅうきゅう", 40 to "よんじゅっ",
            41 to "よんじゅういっ", 42 to "よんじゅうに", 43 to "よんじゅうさん", 44 to "よんじゅうよん",
            45 to "よんじゅうご", 46 to "よんじゅうろっ", 47 to "よんじゅうなな", 48 to "よんじゅうはっ",
            49 to "よんじゅうきゅう", 50 to "ごじゅっ",
            51 to "ごじゅういっ", 52 to "ごじゅうに", 53 to "ごじゅうさん", 54 to "ごじゅうよん",
            55 to "ごじゅうご", 56 to "ごじゅうろっ", 57 to "ごじゅうなな", 58 to "ごじゅうはっ",
            59 to "ごじゅうきゅう"
        )
        val pun = setOf(1,6,8,10,16,18,20,26,28,30,36,38,40,46,48,50,56,58)
        val suffix = if (m in pun) "ぷん" else "ふん"
        return (base[m] ?: "$m") + suffix
    }

    private fun dayReading(d: Int): String {
        val special = mapOf(
            1 to "ついたち", 2 to "ふつか", 3 to "みっか", 4 to "よっか",
            5 to "いつか", 6 to "むいか", 7 to "なのか", 8 to "ようか",
            9 to "ここのか", 10 to "とおか", 14 to "じゅうよっか",
            20 to "はつか", 24 to "にじゅうよっか"
        )
        special[d]?.let { return it }
        val tens = d / 10
        val ones = d % 10
        val onesR = arrayOf("", "いち", "に", "さん", "よん", "ご", "ろく", "しち", "はち", "きゅう")
        var result = when (tens) { 1 -> "じゅう"; 2 -> "にじゅう"; 3 -> "さんじゅう"; else -> "" }
        result += onesR[ones] + "にち"
        return result
    }
}

File: JapaneseClockWidget/JapaneseClockWidget/JapaneseTimeFormatter.swift
import Foundation

enum DisplayMode: String {
    case kanjiOnly = "kanjiOnly"
    case furigana = "furigana"
    case hiraganaOnly = "hiraganaOnly"
}

struct FuriganaPair {
    let text: String
    let reading: String
}

struct JapaneseTimeFormatter {
    private static let dayOfWeekNames = ["日", "月", "火", "水", "木", "金", "土"]
    private static let dayOfWeekReadings = ["にち", "げつ", "か", "すい", "もく", "きん", "ど"]
    private static let amPmNames = ["午前", "午後"]
    private static let amPmReadings = ["ごぜん", "ごご"]

    // Full hiragana readings for hours (1-12)
    private static let hourFullReadings: [Int: String] = [
        1: "いちじ", 2: "にじ", 3: "さんじ", 4: "よじ",
        5: "ごじ", 6: "ろくじ", 7: "しちじ", 8: "はちじ",
        9: "くじ", 10: "じゅうじ", 11: "じゅういちじ", 12: "じゅうにじ"
    ]

    // Full hiragana readings for minutes (1-59)
    private static func minuteFullReading(_ m: Int) -> String {
        let base: [Int: String] = [
            1: "いっ", 2: "に", 3: "さん", 4: "よん",
            5: "ご", 6: "ろっ", 7: "なな", 8: "はっ",
            9: "きゅう", 10: "じゅっ",
            11: "じゅういっ", 12: "じゅうに", 13: "じゅうさん", 14: "じゅうよん",
            15: "じゅうご", 16: "じゅうろっ", 17: "じゅうなな", 18: "じゅうはっ",
            19: "じゅうきゅう", 20: "にじゅっ",
            21: "にじゅういっ", 22: "にじゅうに", 23: "にじゅうさん", 24: "にじゅうよん",
            25: "にじゅうご", 26: "にじゅうろっ", 27: "にじゅうなな", 28: "にじゅうはっ",
            29: "にじゅうきゅう", 30: "さんじゅっ",
            31: "さんじゅういっ", 32: "さんじゅうに", 33: "さんじゅうさん", 34: "さんじゅうよん",
            35: "さんじゅうご", 36: "さんじゅうろっ", 37: "さんじゅうなな", 38: "さんじゅうはっ",
            39: "さんじゅうきゅう", 40: "よんじゅっ",
            41: "よんじゅういっ", 42: "よんじゅうに", 43: "よんじゅうさん", 44: "よんじゅうよん",
            45: "よんじゅうご", 46: "よんじゅうろっ", 47: "よんじゅうなな", 48: "よんじゅうはっ",
            49: "よんじゅうきゅう", 50: "ごじゅっ",
            51: "ごじゅういっ", 52: "ごじゅうに", 53: "ごじゅうさん", 54: "ごじゅうよん",
            55: "ごじゅうご", 56: "ごじゅうろっ", 57: "ごじゅうなな", 58: "ごじゅうはっ",
            59: "ごじゅうきゅう"
        ]
        let suffix: String = {
            switch m {
            case 1, 3, 4, 6, 8, 10,
                 11, 13, 14, 16, 18, 20,
                 21, 23, 24, 26, 28, 30,
                 31, 33, 34, 36, 38, 40,
                 41, 43, 44, 46, 48, 50,
                 51, 53, 54, 56, 58:
                return "ぷん"
            default:
                return "ふん"
            }
        }()
        return (base[m] ?? "\(m)") + suffix
    }

    // Full hiragana readings for months
    private static let monthReadings: [Int: String] = [
        1: "いちがつ", 2: "にがつ", 3: "さんがつ", 4: "しがつ",
        5: "ごがつ", 6: "ろくがつ", 7: "しちがつ", 8: "はちがつ",
        9: "くがつ", 10: "じゅうがつ", 11: "じゅういちがつ", 12: "じゅうにがつ"
    ]

    // Full hiragana readings for days
    private static func dayReading(_ d: Int) -> String {
        let special: [Int: String] = [
            1: "ついたち", 2: "ふつか", 3: "みっか", 4: "よっか",
            5: "いつか", 6: "むいか", 7: "なのか", 8: "ようか",
            9: "ここのか", 10: "とおか", 14: "じゅうよっか",
            20: "はつか", 24: "にじゅうよっか"
        ]
        if let s = special[d] { return s }
        // Regular: じゅういちにち, etc.
        let tens = d / 10
        let ones = d % 10
        var result = ""
        if tens == 1 { result += "じゅう" }
        else if tens == 2 { result += "にじゅう" }
        else if tens == 3 { result += "さんじゅう" }
        let onesReadings = ["", "いち", "に", "さん", "よん", "ご", "ろく", "しち", "はち", "きゅう"]
        result += onesReadings[ones]
        result += "にち"
        return result
    }

    // MARK: - Plain text formatters

    static func formatTime(from date: Date) -> String {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        let minute = calendar.component(.minute, from: date)
        let amPm = hour < 12 ? amPmNames[0] : amPmNames[1]
        let displayHour = hour == 0 ? 12 : (hour > 12 ? hour - 12 : hour)
        if minute == 0 {
            return "\(amPm)\(displayHour)時"
        }
        return "\(amPm)\(displayHour)時\(minute)分"
    }

    static func formatTimeShort(from date: Date) -> String {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        let minute = calendar.component(.minute, from: date)
        let displayHour = hour == 0 ? 12 : (hour > 12 ? hour - 12 : hour)
        if minute == 0 {
            return "\(displayHour)時"
        }
        return "\(displayHour)時\(minute)分"
    }

    static func formatHour(from date: Date) -> String {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        let displayHour = hour == 0 ? 12 : (hour > 12 ? hour - 12 : hour)
        return "\(displayHour)時"
    }

    static func formatMinute(from date: Date) -> String {
        let calendar = Calendar.current
        let minute = calendar.component(.minute, from: date)
        if minute == 0 { return "" }
        return "\(minute)分"
    }

    static func formatAmPm(from date: Date) -> String {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        return hour < 12 ? amPmNames[0] : amPmNames[1]
    }

    static func formatDate(from date: Date) -> String {
        let calendar = Calendar.current
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        let weekday = calendar.component(.weekday, from: date) - 1
        let dayName = dayOfWeekNames[weekday]
        return "\(month)月\(day)日 \(dayName)曜日"
    }

    static func formatDateShort(from date: Date) -> String {
        let calendar = Calendar.current
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        return "\(month)月\(day)日"
    }

    static func formatYear(from date: Date) -> String {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        return "\(year)年"
    }

    // MARK: - Hiragana-only formatters (fully hiragana, no numerals)

    static func formatAmPmHiragana(from date: Date) -> String {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        return hour < 12 ? amPmReadings[0] : amPmReadings[1]
    }

    static func formatHourHiragana(from date: Date) -> String {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        let displayHour = hour == 0 ? 12 : (hour > 12 ? hour - 12 : hour)
        return hourFullReadings[displayHour] ?? "\(displayHour)じ"
    }

    static func formatMinuteHiragana(from date: Date) -> String {
        let calendar = Calendar.current
        let minute = calendar.component(.minute, from: date)
        if minute == 0 { return "" }
        return minuteFullReading(minute)
    }

    static func formatTimeShortHiragana(from date: Date) -> String {
        let hourStr = formatHourHiragana(from: date)
        let minStr = formatMinuteHiragana(from: date)
        return hourStr + minStr
    }

    static func formatDateHiragana(from date: Date) -> String {
        let calendar = Calendar.current
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        let weekday = calendar.component(.weekday, from: date) - 1
        let monthStr = monthReadings[month] ?? "\(month)がつ"
        let dayStr = dayReading(day)
        let weekdayStr = dayOfWeekReadings[weekday] + "ようび"
        return "\(monthStr) \(dayStr) \(weekdayStr)"
    }

    static func formatDateShortHiragana(from date: Date) -> String {
        let calendar = Calendar.current
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        let monthStr = monthReadings[month] ?? "\(month)がつ"
        let dayStr = dayReading(day)
        return "\(monthStr) \(dayStr)"
    }

    static func formatYearHiragana(from date: Date) -> String {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        return numberToHiragana(year) + "ねん"
    }

    // MARK: - Furigana pair formatters

    static func hourPairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        let displayHour = hour == 0 ? 12 : (hour > 12 ? hour - 12 : hour)
        return [
            FuriganaPair(text: "\(displayHour)", reading: ""),
            FuriganaPair(text: "時", reading: "じ")
        ]
    }

    static func minutePairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let minute = calendar.component(.minute, from: date)
        if minute == 0 { return [] }
        return [
            FuriganaPair(text: "\(minute)", reading: ""),
            FuriganaPair(text: "分", reading: minuteSuffix(minute))
        ]
    }

    static func amPmPairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let hour = calendar.component(.hour, from: date)
        let index = hour < 12 ? 0 : 1
        return [
            FuriganaPair(text: amPmNames[index], reading: amPmReadings[index])
        ]
    }

    static func timeShortPairs(from date: Date) -> [FuriganaPair] {
        return hourPairs(from: date) + minutePairs(from: date)
    }

    static func datePairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        let weekday = calendar.component(.weekday, from: date) - 1
        return [
            FuriganaPair(text: "\(month)", reading: ""),
            FuriganaPair(text: "月", reading: "がつ"),
            FuriganaPair(text: "\(day)", reading: ""),
            FuriganaPair(text: "日", reading: "にち"),
            FuriganaPair(text: " ", reading: ""),
            FuriganaPair(text: dayOfWeekNames[weekday], reading: dayOfWeekReadings[weekday]),
            FuriganaPair(text: "曜日", reading: "ようび")
        ]
    }

    static func dateShortPairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        return [
            FuriganaPair(text: "\(month)", reading: ""),
            FuriganaPair(text: "月", reading: "がつ"),
            FuriganaPair(text: "\(day)", reading: ""),
            FuriganaPair(text: "日", reading: "にち")
        ]
    }

    static func yearPairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        return [
            FuriganaPair(text: "\(year)", reading: ""),
            FuriganaPair(text: "年", reading: "ねん")
        ]
    }

    static func weekdayPairs(from date: Date) -> [FuriganaPair] {
        let calendar = Calendar.current
        let weekday = calendar.component(.weekday, from: date) - 1
        return [
            FuriganaPair(text: dayOfWeekNames[weekday], reading: dayOfWeekReadings[weekday]),
            FuriganaPair(text: "曜日", reading: "ようび")
        ]
    }

    // MARK: - Helpers

    private static func minuteSuffix(_ minute: Int) -> String {
        // ぷん: 1, 3, 4, 6, 8, 10 — and their multiples (e.g. 13, 14, 23, 24 ...)
        switch minute {
        case 1, 3, 4, 6, 8, 10,
             11, 13, 14, 16, 18, 20,
             21, 23, 24, 26, 28, 30,
             31, 33, 34, 36, 38, 40,
             41, 43, 44, 46, 48, 50,
             51, 53, 54, 56, 58: return "ぷん"
        default: return "ふん"
        }
    }

    private static func numberToHiragana(_ n: Int) -> String {
        let digits = ["れい", "いち", "に", "さん", "よん", "ご", "ろく", "なな", "はち", "きゅう"]
        if n < 10 { return digits[n] }
        // For years like 2026
        let str = String(n)
        var result = ""
        for ch in str {
            if let d = Int(String(ch)) {
                result += digits[d]
            }
        }
        return result
    }
}
